Vhar 7462 suun turpot analytiikalle

rckopteri opened the pull request:
Uusi rajapinta analytiikkaan.
Palautetaan hakuajan puitteissa turvallisuuspoikkeamat json muodossa.

rckopteri committed:
Tuo XSDsta konvertoitu poikkeama-rest tiedosto json schemaksi.

Tätä muokataan vielä poistamalla numeroenumit ja laittamalla tilalle tekstimuotoiset tiedot.

rckopteri committed:
Uusi kysely hakemaan turvallisuuspoikkeamia ajanjakson sisältä

rckopteri committed:
Lisää uusi endpoint turvallisuuspoikkeamien hakemiseen

rckopteri committed:
Tuo uudet turvallisuuspoikkeamaan liittyvät json skeemat käyttöön

rckopteri committed:
Muokkaa tietokannasta haetut turpot clojure mapiksi, jotta ne voidaan konvertoida jsoniksi

Ja tee niille pari yksinkertaista testiä

rckopteri committed:
Merge branch 'develop' into VHAR-7462-suun-turpot-analytiikalle

rckopteri committed:
Koitetaan erilaisella käyttäjällä.

analytiikka-kayttaja ei ollut lokaalisti muutta ci putkessa oli

rckopteri committed:
Merge branch 'develop' into VHAR-7462-suun-turpot-analytiikalle

# Conflicts:
#	src/clj/harja/palvelin/integraatiot/api/tyokalut/json_skeemat.clj

rckopteri committed:
Päivitä turvallisuuspoikkeaman vastauksen json schemaa

rckopteri committed:
Päivitä turvallisuus poikkeaman vastauksen generoiva koodi vastaamaan json schemaa

rckopteri committed:
Päivitä testi taas toimivaksi

jpluomajoki-solita commented on src/clj/harja/palvelin/integraatiot/api/turvallisuuspoikkeama.clj:
```diff
@@ -264,6 +267,188 @@
+    (poikkeamatoimenpide turvallisuuspoikkeama)
+    (poikkeamaliite turvallisuuspoikkeama)))
+(defn hae-turvallisuuspoikkeamat [db {:keys [alkuaika loppuaika] :as parametrit} kayttaja]
+  (let [_ (println "hae-turvallisuuspoikkeamat :: parametrit" (pr-str parametrit))
```
Ylimääräistä printtiä, ei periaatteessa haittaa tuotannossa.

rckopteri replied:
Nyt toteutettu. Laitoin todo:n ettei se unohdu siltä, joka tämän ikinä loppuun saattaa. Loppuun saattaja oli näimmä minä.

jpluomajoki-solita commented on src/clj/harja/palvelin/integraatiot/api/turvallisuuspoikkeama.clj:
```diff
@@ -264,6 +267,188 @@
+        turpot (turvallisuuspoikkeamat/hae-turvallisuuspoikkeamat-lahetettavaksi-analytiikalle db {:alku (pvm/rajapinta-str-aika->sql-timestamp alkuaika)
+                                                                                              :loppu (pvm/rajapinta-str-aika->sql-timestamp loppuaika)})
+        ;; Konvertoidaan turpot sellaiseen muotoon, että ne voidaan kääntää kutsukäsittelyssä jsoniksi. Tässä vaiheessa ne ovat mäppeineä nimestään huolimatta
+        json-turbot (map
```
turbot -> turpot? Saa mun puolesta jättää näinkin :smile:

rckopteri replied:
hah :) hyvä bongaus. Korjattu

jpluomajoki-solita commented on src/clj/harja/palvelin/integraatiot/api/turvallisuuspoikkeama.clj:
```diff
@@ -280,19 +465,30 @@
+    (julkaise-reitti
+      http :hae-turvallisuuspoikkeamat
+      (GET "/api/turvallisuuspoikkeamat/:alkuaika/:loppuaika" request
+        ;; TODO: HOX!!!! ihan väärä integraatio tallennetaan kantaan
```
Pitäisikö tämä vielä korjata?

rckopteri replied:
Jotettu oikealle toteuttajalle toteutettavaksi. Tämä oli vain suunnitelma, ei toteutus.

jpluomajoki-solita reviewed (commented):
Vaikuttaisi hyvältä, yksi korjattavan näköinen todo-kommentti pisti silmään

rckopteri committed:
Tee tarkemmat validoinnit turpojen hakuun

rckopteri committed:
Mahdollista skeemassa useamman vamman lähettäminen

rckopteri committed:
Muuta numerot teksteiksi

rckopteri committed:
Lisää pikkukuvan bytet base64 enkoodattuna jsonille

rckopteri committed:
Tarkista analytiikalle lähetettävän turvallisuuspoikkeaman tietojen oikeellisuus

tietoja tarkkailevalla yksikkötestillä.
Avuksi on tehty funktio, jonka avulla voidaan vertailla lähes samantyyppisiä tekstejä.

rckopteri committed:
Merge branch 'develop' into VHAR-7462-suun-turpot-analytiikalle

rckopteri committed:
Siirrä turvallisuuspoikkeamien analytiikan haku analytiikan alle.

ja täydennä samalla hieman testejä.

rckopteri committed:
Typo turbot -> turpot

jpluomajoki-solita commented on resources/api/schemas/turvallisuuspoikkeama-vastaus.schema.json:
```diff
@@ -812,146 +513,86 @@
-                                    ],
-                                    "description": "Poikkeaman tyyppi. Syötä ilmoitusta koskevan poikkeaman tyyppi (voit valita useita vaihtoehtoja). 8 = Työtapaturma 16 = Muuu turvallisuuspoikkeama 32 = Vaaratilanne 64 = Turvallisuushavainto",
+                                    "type": "string",
+                                    "description": "Poikkeaman tyyppi. Syötä ilmoitusta koskevan poikkeaman tyyppi (voit valita useita vaihtoehtoja). 8 = Työtapaturma 16 = Muu turvallisuuspoikkeama 32 = Vaaratilanne 64 = Turvallisuushavainto",
```
Description ei enää vastaa enumia, muutama muukin muuttunut intistä stringiksi

jpluomajoki-solita commented on resources/api/schemas/turvallisuuspoikkeama-vastaus.schema.json:
```diff
@@ -244,203 +138,168 @@
-                                        "$ref": "#/definitions/vammanlaatuType"
-                                    }
-                                ]
+                                "description": "Aiheutuneen vamman laatu Tämä kenttä on käytössä, jos tyyppi-kentän arvo on 4. Otetaan poikkeuksellisesti array muodossa vammat mukaan.",
```
Tyyppi-kenttä on muuttunut string-enumiksi, pitäisikö dokumentointi päivittää?  Sama homma muutamassa muussa kentässä

jpluomajoki-solita reviewed (approved):
dokumentaatio ehkä kaipaa vielä päivitystä, muuten näyttää hyvältä

rckopteri commented on test/clj/harja/palvelin/integraatiot/api/analytiikka_test.clj:
```diff
@@ -25,12 +20,40 @@
+        _ (u (format "DELETE FROM korjaavatoimenpide WHERE turvallisuuspoikkeama = %s", id))
+        _ (u (format "DELETE FROM turvallisuuspoikkeama WHERE id = %s", id))]))
+
+(defn valiaikainen-kayttaja [kayttajanimi]
```
Harkitseppa käyttäjän lisäämistä ihan oikeasti, koska se helpottaisi käsin testaamista.

rckopteri commented on test/clj/harja/palvelin/integraatiot/api/analytiikka_test.clj:
```diff
@@ -163,3 +182,196 @@
     (is (= alkup-maara lopullinen-maara))
     (is (= uusi-maara muokattu-maara))))
+
+(deftest hae-turvallisuuspoikkeamat-analytiikalle-ei-kayttoikeutta
```
Lisää testi, joka käyttää ihan oikeaa käyttäjää, jolla ei ole oikeutta.

rckopteri committed:
Merge branch 'develop' into VHAR-7462-suun-turpot-analytiikalle

# Conflicts:
#	src/clj/harja/palvelin/integraatiot/api/turvallisuuspoikkeama.clj
#	test/clj/harja/palvelin/integraatiot/api/turvallisuuspoikkeaman_kirjaus_test.clj

rckopteri committed:
Korjaa katselmointikommentit

rckopteri committed:
Turisanomat on poistettu, joten siirrä sieltä koodit vain analytiikan sanomien käyttöön

rckopteri committed:
Poista turi analytiikkatesteistä

rckopteri committed:
Siirrä tekstien vertailu kaikkien testien käyttöön.

rckopteri committed:
Poista ylimääräiset analytiikka-testerin luonnit

rckopteri committed:
Rajoita analytiikka apin suunniteltujen arvojen palautukset koskemaan vain urakan voimassaolevia vuosia

rckopteri committed:
Palauta virheviesti samalle vuodelle kuin oli aiemmin
